mt_gateway_test.go

```diff
@@ -0,0 +1,333 @@
+package main
+
+import (
+	"context"
+	b64 "encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/http/httptest"
+	"os"
+	"os/signal"
+	"strconv"
+	"strings"
+	"syscall"
+	"testing"
+	"time"
+
+	"github.com/gorilla/mux"
+	"github.com/ory/dockertest/v3"
+	"github.com/phayes/freeport"
+	"github.com/stretchr/testify/require"
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/rudderlabs/rudder-server/app"
+	"github.com/rudderlabs/rudder-server/testhelper"
+	"github.com/rudderlabs/rudder-server/testhelper/destination"
+	thEtcd "github.com/rudderlabs/rudder-server/testhelper/etcd"
+	"github.com/rudderlabs/rudder-server/testhelper/health"
+	"github.com/rudderlabs/rudder-server/testhelper/rand"
+	whUtil "github.com/rudderlabs/rudder-server/testhelper/webhook"
+	"github.com/rudderlabs/rudder-server/utils/types/deployment"
+)
+
+func TestMultiTenantGateway(t *testing.T) {
+	ctx, cancel := signal.NotifyContext(context.Background(), syscall.SIGTERM, syscall.SIGINT)
+	defer cancel()
+
+	pool, err := dockertest.NewPool("")
+	require.NoError(t, err)
+
+	var (
+		group             errgroup.Group
+		etcdContainer     *thEtcd.Resource
+		postgresContainer *destination.PostgresResource
+		serverInstanceID  = "1"
+	)
+
+	group.Go(func() (err error) {
+		postgresContainer, err = destination.SetupPostgres(pool, t)
+		return err
+	})
+	group.Go(func() (err error) {
+		etcdContainer, err = thEtcd.Setup(pool, t)
+		return err
+	})
+	require.NoError(t, group.Wait())
+
+	webhook := whUtil.NewRecorder()
+	t.Cleanup(webhook.Close)
+
+	writeKey := rand.String(27)
+	workspaceID := rand.String(27)
+	marshalledWorkspaces := testhelper.FillTemplateAndReturn(t, "testdata/mtGatewayTest01.json", map[string]string{
+		"writeKey":    writeKey,
+		"workspaceId": workspaceID,
+		"webhookUrl":  webhook.Server.URL,
+	})
+	require.NoError(t, err)
+
+	backendConfRouter := mux.NewRouter()
+	if testing.Verbose() {
+		backendConfRouter.Use(func(next http.Handler) http.Handler {
+			return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				t.Logf("BackendConfig server call: %+v", r)
+				next.ServeHTTP(w, r)
+			})
+		})
+	}
+	backendConfRouter.HandleFunc("/hostedWorkspaceConfig", func(w http.ResponseWriter, r *http.Request) {
+		n, err := w.Write(marshalledWorkspaces.Bytes())
+		require.NoError(t, err)
+		require.Equal(t, marshalledWorkspaces.Len(), n)
+	}).Methods("GET")
+
+	backendConfigSrv := httptest.NewServer(backendConfRouter)
+	t.Logf("BackendConfig server listening on: %s", backendConfigSrv.URL)
+	t.Cleanup(backendConfigSrv.Close)
+
+	httpPort, err := freeport.GetFreePort()
+	require.NoError(t, err)
+	httpAdminPort, err := freeport.GetFreePort()
```

Comment:
Is there no chance that both `httpPort` and `httpAdminPort` end up being same?

Reply:
good point/catch, free port will close the port so it can be used again. So there is no guarantee this won't return the same port.

Reply:
Yes because we should use `httpPort` first and then ask for another one but there is very little chance that it would happen, and if it happens it'll just make the test fail. The alternative is to do a helper around it. I'll have a stab see if I can do a fix because they are both used by RS later on so we can't really use one and then ask for another. We can only keep on asking until we get a different one.